Sources/SPMBuildCore/PluginScriptRunner.swift

```diff
@@ -61,18 +62,33 @@ public protocol PluginScriptRunner {
     var hostTriple: Triple { get }
 }
 
-/// Protocol by which `PluginScriptRunner.runPluginScript()` communicates back to the caller.
+/// Protocol by which `PluginScriptRunner` communicates back to the caller as it compiles and runs plugins.
 public protocol PluginScriptRunnerDelegate {
+    /// Called immediately before compiling a plugin. Will not be called if the plugin didn't have to be compiled. This call is always followed by a `didCompilePlugin()` but is mutually exclusive with a `skippedCompilingPlugin()` call.
+    func willCompilePlugin(commandLine: [String], environment: EnvironmentVariables)
+    
+    /// Called immediately after compiling a plugin (regardless of whether it succeeded or failed). Will not be called if the plugin didn't have to be compiled. This call is always follows a `willCompilePlugin()` but is mutually exclusive with a `skippedCompilingPlugin()` call.
+    func didCompilePlugin(result: PluginCompilationResult)
+    
+    /// Called if a plugin didn't need to be compiled because previous compilation results were still valid. In this case neither `willCompilePlugin()` nor `didCompilePlugin()` will be called.
+    func skippedCompilingPlugin(cachedResult: PluginCompilationResult)
+    
     /// Called for each piece of textual output data emitted by the plugin. Note that there is no guarantee that the data begins and ends on a UTF-8 byte sequence boundary (much less on a line boundary) so the delegate should buffer partial data as appropriate.
     func handleOutput(data: Data)
     
     /// Called for each length-delimited message received from the plugin. The `responder` is closure that can be used to send one or more messages in reply.
     func handleMessage(data: Data, responder: @escaping (Data) -> Void) throws
 }
-
+public extension PluginScriptRunnerDelegate {
```

Comment:
does this extension need to be public?

Reply:
I thought it did, but I just tried now and it doesn't seem it does — thanks for calling that out.  I have fixed it.

Reply:
It does if we want to allow the delegate methods to be optional.  We could also require clients to implement them, thus getting rid of the extension.  Do you have a preference?

Reply:
I generally prefer to make that an explicit decision of the client, this kind of extension will hide it from them

Reply:
The reason I added the defaults was because not all the callbacks make sense for all the use cases, e.g. if I chose to only compile the plugin, then having to implement the callbacks for running the plugin don't make sense.

That's probably an argument for splitting out the compiling and running into separate calls and separate delegate protocols.  I started down that road but it turned into a bit more of a change than I thought was warranted.  But maybe that's the right thing to do even in this PR (as opposed to a follow-on fix).

Reply:
OK:  I split this out into two separate protocols and use composition for the runner delegate, which makes things clearer.  I also got rid of the extensions providing default implementations, since SwiftPM has been moving away from those.